test/cuda/utils.cu

```diff
@@ -0,0 +1,180 @@
+#include "../utils.hh"
+
+#if defined(BLAS_HAVE_CUBLAS)
+    #include <cuda_fp16.h>
+#elif defined(BLAS_HAVE_ROCBLAS)
+    #include <hip/hip_fp16.h>
+#endif
+
+//------------------------------------------------------------------------------
+/// @return ceil( x / y ), for integer type T.
+template <typename T>
+inline constexpr T ceildiv( T x, T y )
+{
+    return T( (x + y - 1) / y );
+}
+
+//==============================================================================
+// Overloads to enable templating.
+
+//------------------------------------------------------------------------------
+// Template implementation when C++ has default rules for conversion
+// (or no conversion needed).
+template <typename src_t, typename dst_t>
+__host__ __device__
+inline void copy_scalar( src_t src, dst_t& dst )
+{
+    dst = dst_t( src );
+}
+
+//------------------------------------------------------------------------------
+// Overloaded implementations for specific cases.
+__host__ __device__
+inline void copy_scalar( float src, __half& dst )
+{
+    dst = __float2half( src );
+}
+
+//------------------------------------------------------------------------------
+__host__ __device__
+inline void copy_scalar( double src, __half& dst )
+{
+#ifdef __NVCC__
+    dst = __double2half( src );
+#else
+    dst = __float2half( (float)src );
+#endif
+}
+
+//------------------------------------------------------------------------------
+__host__ __device__
+inline void copy_scalar( __half src, float& dst )
+{
+    dst = __half2float( src );
+}
+
+//------------------------------------------------------------------------------
+__host__ __device__
+inline void copy_scalar( __half src, double& dst )
+{
+    // no __half2double, so do in 2 steps
+    dst = double( __half2float( src ) );
+}
+
+//==============================================================================
+// GPU function.
+const int blk_x = 32;
+const int blk_y = 4;
+
+//------------------------------------------------------------------------------
+// GPU device routine, called from GPU kernel.
+//
+// Each thread-block does a blk_x by blk_y block of the matrix;
+// each thread does 1 entry in the block.
+// Because of the CUDA max y grid dimension of 65535, the entire grid is
+// repeated in the y dimension with step = gridDim.y * blockDim.y,
+// so thread (i, j) will do entries (i, j), (i, j + step), (i, j + 2*step), ...
+// The max x grid dimension is 2^31-1, so there's no need to repeat.
+// Cf. magma/magmablas/slag2h.cu
+//
+template <typename src_t, typename dst_t>
+__device__
+void copy_matrix_device(
+    int m, int n,
```

Comment:
`int64_t` for dimensions (`m, n, ld, i, j`) in all these routines. Otherwise it runs into problems with a 46342 x 46342 matrix, where j*ld can overflow.